### PR TITLE
Implement Interpretation rule engine

### PR DESCRIPTION
**Pull request by kstam**

### CODE REVIEW

File: src/domain/model/test/interpretation/Condition.test.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        property: 'c',
+        comparator: '==',
+        value: false,
+      });
+
+      expect(condition.evaluate({ c: true })).toBe(false);
+      expect(condition.evaluate({ c: false })).toBe(true);
+    });
+
+    it('correctly evaluates "1=" for numbers', () => {
```
Comment (StevePole) [inline src/domain/model/test/interpretation/Condition.test.ts:87]: typo

File: src/domain/model/test/interpretation/InterpretationRules.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        },
+      },
+    ]);
+
+    const results = new Results(new UserId(), { c: 3 }, ConfidenceLevel.LOW);
+    const interpretations = interpretationRule.interpret(results);
+
+    expect(interpretations[0]?.name).toEqual('Some pattern {{value}}');
+    expect(interpretations[0]?.theme).toEqual(InterpretationTheme.NEGATIVE);
+    expect(interpretations[0]?.variables.get('value')).toEqual(3);
```
Comment (StevePole) [inline src/domain/model/test/interpretation/InterpretationRules.test.ts:31]: So you're going to have the client substitute the value, or did I misunderstand?

File: src/domain/model/test/interpretation/InterpretationRules.test.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  });
+
+  it('produces a null result when the condition evaluates to false', () => {
```
Comment (StevePole) [inline src/domain/model/test/interpretation/InterpretationRules.test.ts:34]: I would change the description, empty array isn't null.

File: src/domain/model/test/interpretation/InterpretationRules.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import { Interpretation } from './Interpretation';
+import { DomainValidationError } from '../../DomainValidationError';
+
+export class InterpretationRules {
+  constructor(private rules: InterpretationRule[]) {}
+
+  interpret(results: Results): Array<Interpretation> {
+    return this.rules
+      .map((rule) => rule.interpret(results))
+      .filter((interpretation) => interpretation !== null) as Interpretation[];
```
Comment (StevePole) [inline src/domain/model/test/interpretation/InterpretationRules.ts:13]: Here you are expecting null, slightly confused by that...

File: src/domain/model/test/interpretation/InterpretationRules.test.ts (hunk 2 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.test.ts:31]: Yes. the reason is to allow for localisation in the future.

File: src/domain/model/test/interpretation/InterpretationRules.test.ts (hunk 3 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.test.ts:34]: oh yeah.. I changed things around and forgot to rename. good catch

File: src/domain/model/test/interpretation/InterpretationRules.ts (hunk 4 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.ts:13]: InterpretationRule.interpret returns `null` if the rule doesn't match.

File: src/domain/model/test/interpretation/InterpretationRules.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  interpret(results: Results): Interpretation | null {
+    if (this.condition.evaluate(results.details)) {
+      return this.output.evaluate(results);
+    }
+    return null;
+  }
+
+  static from(interpretationRuleSchema: any) {
+    return new InterpretationRule(
```
Comment (StevePole) [inline src/domain/model/test/interpretation/InterpretationRules.ts:35]: Do you need validation in here, or is that handled somewhere else?

File: src/domain/model/test/interpretation/InterpretationRules.ts (hunk 4 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.ts:13]: I can split to `matches()` and `getInterpretation()`. Probably would read better

File: src/domain/model/test/interpretation/InterpretationRules.ts (hunk 5 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.ts:35]: Yeah.. could add. There will be an error thrown from the OutputPattern or the Condition based on what is invalid but better to have meaningful errors

File: src/domain/model/test/interpretation/InterpretationRules.ts (hunk 4 above)
Comment (StevePole) [reply src/domain/model/test/interpretation/InterpretationRules.ts:13]: It's fairly clear, I was just confused as I had been looking at InterpretationRules, which returns an empty array.

File: src/domain/model/test/interpretation/InterpretationRules.ts (hunk 5 above)
Comment (kstam) [reply src/domain/model/test/interpretation/InterpretationRules.ts:38]: Added 👌

**Pull request merged**